rules/directory/providers.bzl

```diff
@@ -0,0 +1,46 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
```

Comment:
Providers should be under `./private` directory, to make it clear, it's an implementation detail of directory rules. Probably also utils.bzl

Reply:
The providers are the API for this feature. Otherwise, we have no way of writing:
```
cc_toolchain = rule(
  attrs = {
    "sysroot": attr.label(providers = [DirectoryInfo]),
  }
)
```